src/train_cmd.cpp

```diff
@@ -3178,19 +3179,25 @@ bool TrainController(Train *v, Vehicle *nomove, bool reverse)
 						/* In front of a red signal */
 						Trackdir i = FindFirstTrackdir(trackdirbits);
 
-						/* Don't handle stuck trains here. */
+						/* Don't handle trains waiting at path signals here. */
 						if (HasBit(v->flags, VRF_TRAIN_STUCK)) return false;
 
 						if (!HasSignalOnTrackdir(gp.new_tile, ReverseTrackdir(i))) {
 							v->cur_speed = 0;
 							v->subspeed = 0;
 							v->progress = 255; // make sure that every bit of acceleration will hit the signal again, so speed stays 0.
-							if (!_settings_game.pf.reverse_at_signals || ++v->wait_counter < _settings_game.pf.wait_oneway_signal * DAY_TICKS * 2) return false;
+							++v->wait_counter;
+							/* Note that wait_counter effectively increments at 2 per tick at block signals */
+							if (v->wait_counter % 64 == 0) v->AddCargoPotentialPenalty(32);
```

Comment:
64? 32?

Reply:
`wait_counter` increments at a rate of 2 per in-game tick, so when it reaches 64 the vehicle has been waiting for 32 ticks.